doc/book/box/indexes.rst

```diff
@@ -24,46 +20,454 @@ The first index defined on a space is called the **primary key index**,
+        *   - Feature
+            - TREE
+            - HASH
+            - RTREE
+            - BITSET
+
+        *   - unique
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - non-unique
+            - \+
+            - \-
+            - \+
+            - \+
+
+        *   - :ref:`is_nullable <box_space-is_nullable>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - multi-part
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - :ref:`multikey <box_space-path_multikey>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - :ref:`partial-key search <partial_key_search>`
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - can be primary key
+            - \+
+            - \+
+            - \-
+            - \-
+
+        *   - exclude_null (version 2.8+)
+            - \+
+            - \-
+            - \-
+            - \-
+
+        *   - :doc:`iterator types </reference/reference_lua/box_index/pairs>`
+            - ALL, EQ, REQ, GT, GE, LT, LE
+            - ALL, EQ, GT
+            - ALL, EQ, GT, GE, LT, LE, OVERLAPS, NEIGHBOR
+            - ALL, EQ, BITS_ALL_SET, BITS_ANY_SET, BITS_ALL_NOT_SET
+
+********************************************************************************
+TREE indexes
+********************************************************************************
+
+The default index type is 'TREE'.
+TREE indexes are provided by memtx and vinyl engines, can index unique and
+non-unique values, support partial key searches, comparisons and ordered results.
+
+This is a universal type of indexes, for most cases it will be the best choice.
+
+Additionally, memtx engine supports HASH, RTREE and BITSET indexes.
+
+********************************************************************************
+HASH indexes
+********************************************************************************
+
+HASH indexes require unique fields and loses to TREE in almost all respects.
+So we do not recommend to use it in the applications.
+HASH is now present in Tarantool mainly because of backward compatibility.
+
+Here are some tips. Do not use HASH index:
+
+* just if you want to
+* if you think that HASH is faster with no performance metering
+* if you want to iterate over the data
+* for primary key
+* as an only index
+
+Use HASH index:
+
+* if it is a secondary key
+* if you 100% won't need to make it non-unique
+* if you really need that 2-5% performance improvement
+* if you have taken measurements on your data and you see an increase in performance
+* if you save every byte on tuples (HASH is a little more compact)
+
+********************************************************************************
+RTREE indexes
+********************************************************************************
+
+RTREE is a multidimensional index supporting up to 32 dimensions.
```

Comment:
AFAIK "up to 20" dimensions - https://github.com/tarantool/tarantool/blob/f8449727a02e588f6455d937089a96ed3bcf194f/src/lib/salad/rtree.h#L71